### PR TITLE
Add window with a list of all keybindings

### LINKED ISSUE #3314: Add a window with a list of all keybindings
To show all key bindings in one place, maybe to allow redefining them as well.

### PR DESCRIPTION
**Pull request by PremadeS**
`ShortcutOptionsWidget` to show list of keybindings in Edit->Preferences + `ShortcutManager` class to query shortcuts

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://cutter.re/docs/contributing/code/getting-started.html) to this repository
- [x] I made sure to follow the project's [coding style](https://cutter.re/docs/contributing/code/development-guidelines.html)
- [ ] I've updated the [documentation](https://cutter.re/docs/user-docs.html) with the relevant information (if needed)

**Detailed description**

Adds a window with a list of all keybindings/shortcuts in `Edit->Preferences`
window is named `ShortcutOptionsWidget` and contains 
* `Command` - unique name for each shortcut in the format <context.function> where context is **Graph**, **General** etc.
* `Shorcut`
* `Description`
* `QLineEdit` to filter through shortcuts, filter scans all three of the above defined columns

Default shortcuts are stored in `DefaultShortcuts.cpp`
`ShortcutManager` checks for custom keySequences in `settings` if none are present for a specific shortcut then default keySequence is returned.
Widgets query `ShortcutManager` using a unique name for each shortcut. 
A shortcut can have multiple keySequences so a stringlist is stored in `settings` instead of a single string for each shortcut

shortcuts-pr

Note: Cutter [shortcuts docs](https://cutter.re/docs/user-docs/shortcuts.html) are really outdated

**Test plan (required)**

Go to `Edit->Preferences` open `Shortcuts` option and check all of the shortcuts. Try different ones to see if they work correctly + try filtering

**Closing issues**

closes #3314

### CODE REVIEW

**Comment by karliss** [comment]
It will take a few days until I have time to check it out properly. But from a first glance the current concept of `ActionManager` needs to be thrown in the bin and redone from scratch. You can't add pointers of widget owned actions to a central action manager like that. It's wrong both from memory management/object lifetime perspective and usage perspective (I am guessing that you won't see  all the shortcuts until the corresponding widget is created).

Sure there will be some kind of central manager, which the shortcut window can query, but how it interacts with rest of widgets needs to be reconsidered.

**Comment by PremadeS** [comment]
> It will take a few days until I have time to check it out properly. But from a first glance the current concept of `ActionManager` needs to be thrown in the bin and redone from scratch. You can't add pointers of widget owned actions to a central action manager like that. It's wrong both from memory management/object lifetime perspective and usage perspective (I am guessing that you won't see  all the shortcuts until the corresponding widget is created).
> 
> Sure there will be some kind of central manager, which the shortcut window can query, but how it interacts with rest of widgets needs to be reconsidered. 

Understood, what could be a better approach for it? Just a rough idea on how the central widget can work with other widgets.

**Comment by PremadeS** [comment]
@karliss I changed the approach a bit. Let me know if this works better

**Comment by karliss** [comment]
This looks like a step in right direction. Is the code ready for more detailed review, or did you just want a quick confirmation for appraoch before further work?

**Comment by PremadeS** [comment]
> This looks like a step in right direction. Is the code ready for more detailed review, or did you just want a quick confirmation for appraoch before further work?

Ready for review.

However I do have a question, a shortcut can have multiple keySequences, this is easy in `QActions` but in `QShortcut` to set multiple keySequences `setKeys()` is used which was introduced in Qt6 and there is no default constructor for it. So I'm still unsure whether to use it or not.

This is why `getKeySequence()` in `ShortcutManager` returns the first keySequence from the list of potentially many keySequences.

**Comment by PremadeS** [comment]
> This looks like a step in right direction. Is the code ready for more detailed review, or did you just want a quick confirmation for appraoch before further work?

Ready for review

**Review by wargio — COMMENTED** [review]
it seems that is missing a way to update the shortcut when there is an update (besides closing and reopening cutter)

For me this is the perfect job of a QT signal.

**Comment by PremadeS** [comment]
> it seems that is missing a way to update the shortcut when there is an update (besides closing and reopening cutter)
> 
> For me this is the perfect job of a QT signal.

That should be implemented when developing custom shortcut support right? This PR is only for creating a unified window to show all of the default shortcuts

Although I can implement it if needed

File: src/dialogs/preferences/ShortcutOptionsWidget.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        "Console",     "Hex",        "Debug",           "Functions", "Types",   "RegisterRefs",
+        "Threads",     "Omnibar",    "Globals",         "Flags",     "VTables", "ListDock",
+        "Processes",   "Exports",    "Imports"
+    };
+
+    QTreeWidgetItem *globalItem = createCategoryItem(tr("General"));
+
+    QHash<QString, QTreeWidgetItem *> prefixToItem;
+    for (const QString &key : categoryKeys) {
+        prefixToItem[key] = createCategoryItem(tr(qPrintable(key)));
```
Comment (karliss) [inline src/dialogs/preferences/ShortcutOptionsWidget.cpp]: That's not how qt translation system works.

File: src/menus/DisassemblyContextMenu.cpp [hunk 2]
```diff
@@ -637,92 +641,92 @@ void DisassemblyContextMenu::aboutToHideSlot()
 
 QKeySequence DisassemblyContextMenu::getCopySequence() const
 {
-    return QKeySequence::Copy;
+    return Shortcuts()->getKeySequence("Disassembly.copy");
 }
```
Comment (karliss) [inline src/menus/DisassemblyContextMenu.cpp]: Unless any of dissassembly context menu subclasses are overriding these it seems redundant. You can probably get rid of all these methods.

File: src/shortcuts/DefaultShortcuts.cpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
+#include "DefaultShortcuts.h"
+
+const QHash<QString, Shortcut> &getDefaultShortcuts()
+{
+    static const QHash<QString, Shortcut> defaultShortcuts = {
+
+        // General
+        { "focusConsole", { { Qt::Key_Period }, "Focus Console Input" } },
```
Comment (karliss) [inline src/shortcuts/DefaultShortcuts.cpp]: Translation handling.  See some of my recent commits with QT_TRANSLATE_NOOP. Also pay attention to context value so that they match with all the previous action labels from which you copied these.

File: src/shortcuts/ShortcutManager.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+{
+    const auto &defaultShortcuts = getDefaultShortcuts();
+    return tr(qPrintable(defaultShortcuts.value(name).text));
+}
+
+Shortcut ShortcutManager::getShortcut(const QString &name)
+{
+    QList<QKeySequence> ksq = getKeySequences(name);
+    QString text = getText(name);
+    return { ksq, text };
```
Comment (karliss) [inline src/shortcuts/ShortcutManager.cpp]: Proper shortcut remapping system will likely need a bit more metadata for each shortcut description. The approach of gluing key sequence + string will get messy once there are more fields.

File: src/widgets/BreakpointWidget.cpp [hunk 5]
```diff
@@ -193,14 +194,16 @@ BreakpointWidget::BreakpointWidget(MainWindow *main)
...
+    actionDelBreakpoint->setShortcuts(shortcutDelBreakpoint.keySequences);
     actionDelBreakpoint->setShortcutContext(Qt::WidgetShortcut);
     connect(actionDelBreakpoint, &QAction::triggered, this, &BreakpointWidget::delBreakpoint);
     ui->breakpointTreeView->addAction(actionDelBreakpoint);
 
-    actionToggleBreakpoint = new QAction(tr("Toggle breakpoint"), this);
-    actionToggleBreakpoint->setShortcut(Qt::Key_Space);
+    Shortcut shortcutToggleBreakpoint = Shortcuts()->getShortcut("Breakpoint.toggleBreakpoint");
+    actionToggleBreakpoint = new QAction(shortcutToggleBreakpoint.text, this);
+    actionToggleBreakpoint->setShortcuts(shortcutToggleBreakpoint.keySequences);
```
Comment (karliss) [inline src/widgets/BreakpointWidget.cpp]: Might be useful to create an action factory function something like `QAction* ShortcutManager::makeAction(QString id, QWidget* parent)` and `ShortcutManger::setupAction(QAction& action, QString id)`.  It would potentially simplify large fraction of setup calls, allow extending amount of metadata stored in the shortcut description struct, and would potentially give a point where you can setup singal listener for updating the shortcuts.

**Review by karliss — CHANGES_REQUESTED** [review]

**Comment by karliss** [comment]
> That should be implemented when developing custom shortcut support right? This PR is only for creating a unified window to show all of the default shortcuts

I somewhat agree with this one. Let's not try to do everything in single PR. There will be plenty of work within this PR and and the second part. That said during review of this I will be making comments with future remapping in mind so that the the second PR doesn't have to redo all changes touching shortcut handling all over the Cutter.

File: src/shortcuts/DefaultShortcuts.cpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        { "gotoEntry", { { Qt::Key_G }, "Go to Entry" } },
+        { "seek", { { Qt::Key_S }, "Seek" } },
+        { "seekToFunctionEnd", { { Qt::Key_Dollar }, "Seek to Function End" } },
+        { "seekToFunctionStart", { { Qt::Key_AsciiCircum }, "Seek to Function Start" } },
+        { "refreshContens", { { Qt::CTRL | Qt::Key_R }, "Refresh Contents" } },
+        { "back", { { QKeySequence::Back }, "Undo Seek" } },
+        { "forward", { { QKeySequence::Forward }, "Redo Seek" } },
+        { "zoomIn", { { Qt::Key_Plus }, "Zoom In" } },
+        { "zoomOut", { { Qt::Key_Minus }, "Zoom Out" } },
```
Comment (karliss) [inline src/shortcuts/DefaultShortcuts.cpp]: For the sake of consistency might be better to create an explicit category for common/general shortcuts.

File: src/shortcuts/DefaultShortcuts.cpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        { "Strings.showFilter", { { QKeySequence::Find }, "Show Filter" } },
+        { "Strings.clearFilter", { { Qt::Key_Escape }, "Clear Filter" } },
+
+        // Threads
+        { "Threads.showFilter", { { QKeySequence::Find }, "Show Filter" } },
+        { "Threads.clearFilter", { { Qt::Key_Escape }, "Clear Filter" } },
+
+        // Types
+        { "Types.showFilter", { { QKeySequence::Find }, "Show Filter" } },
+        { "Types.clearFilter", { { Qt::Key_Escape }, "Clear Filter" } },
```
Comment (karliss) [inline src/shortcuts/DefaultShortcuts.cpp]: This looks like common behavior, each table widget shouldn't have it's own shortcut for this.

**Review by karliss — CHANGES_REQUESTED** [review]

File: src/shortcuts/DefaultShortcuts.cpp [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    static const QHash<QString, Shortcut> defaultShortcuts = {
+
+        // General
+        { "focusConsole", { { Qt::Key_Period }, "Focus Console Input" } },
+        { "gotoEntry", { { Qt::Key_G }, "Go to Entry" } },
+        { "seek", { { Qt::Key_S }, "Seek" } },
+        { "seekToFunctionEnd", { { Qt::Key_Dollar }, "Seek to Function End" } },
+        { "seekToFunctionStart", { { Qt::Key_AsciiCircum }, "Seek to Function Start" } },
+        { "refreshContens", { { Qt::CTRL | Qt::Key_R }, "Refresh Contents" } },
+        { "back", { { QKeySequence::Back }, "Undo Seek" } },
```
Comment (karliss) [inline src/shortcuts/DefaultShortcuts.cpp]: Single `QKeySequence::StandardKey` can map to multiple `QKeySequences`.  Setting shortcuts for QAction using `StandardKey` uses them all, but creating `QKeySequence` from `StandardKey` will only use one of them.

**Review by karliss — CHANGES_REQUESTED** [review]

**Comment by wargio** [comment]
> This PR is only for creating a unified window to show all of the default shortcuts

oh, i missed this specific point. then ignore my comment :)

**Comment by PremadeS** [comment]
@karliss Just a quick note: since `showFilter` and `clearFilter` are now in the `General` category, I wasn’t sure which context to assign to them in `DefaultShortcuts`. For now, I’ve placed them under `MainWindow`.

Also, some widgets don’t provide text or descriptions for their shortcuts/actions, so there are no corresponding translations in the `ShortcutOptionsWidget`. You probably already knew that, but I thought it was worth mentioning just in case.

File: src/shortcuts/ShortcutManager.cpp (hunk 4 above)
Comment (karliss) [reply src/shortcuts/ShortcutManager.cpp]: The problem still remains, you are still explicitly touching every field in Shortcut even though only one of them is changing.  If you are going to copy the whole struct overriding only one of them do so. That is get the whole default Shortcut struct copy it (this is C++ default copy constructor should be just fine) and replace single field. 

Also get rid of `ShortcutManager::getX(id)` methods except maybe the one for key sequence which needs special handling. Not only it's unnecessary boilerplate for accessing every field, it encourage poor code like what you currently have:
```
s.text = getText(id);
s.context = getContext(id);
...
s.propN = getPropN(id);
```
If you think of what it does it's essentially:
```
shortcut = defaultMap.get(id);
result.text = shortcut.text;
shortcut = defaultMap.get(id);
result.context = shortcut.context;
...
shortcut = defaultMap.get(id);
result.propN = shortcut.propN;
```

When you could have just done something like:
```
result = defaultMap.get(id);
customShortcuts = getCustomShortcuts(id);
if (customShortcuts) {
    result.keySequences = customShorcuts;
}
```

File: src/shortcuts/DefaultShortcuts.cpp (hunk 8 above)
Comment (karliss) [reply src/shortcuts/DefaultShortcuts.cpp]: Did you do anything for addressing this?

File: src/menus/DisassemblyContextMenu.cpp [hunk 9]
```diff
@@ -1149,14 +1063,13 @@ void DisassemblyContextMenu::initAction(QAction *action, QString name, const cha
...
+void DisassemblyContextMenu::initAction(QAction *action, const Shortcut &shortcut, const char *slot)
 {
-    initAction(action, name, slot);
-    if (keySequenceList.empty()) {
+    initAction(action, tr(shortcut.text), slot);
+    if (shortcut.keySequences.empty()) {
         return;
     }
-    action->setShortcuts(keySequenceList);
+    action->setShortcuts(shortcut.keySequences);
```
Comment (karliss) [inline src/menus/DisassemblyContextMenu.cpp]: Can't you use `ShortcutManager::setupAction` here?

File: src/widgets/DebugActions.cpp [hunk 10]
```diff
@@ -57,26 +58,26 @@ DebugActions::DebugActions(QToolBar *toolBar, MainWindow *main) : QObject(main),
 
     // define actions
     actionStart = new QAction(startDebugIcon, startDebugLabel, this);
-    actionStart->setShortcut(QKeySequence(Qt::Key_F9));
+    actionStart->setShortcuts(Shortcuts()->getKeySequences("Debug.start"));
```
Comment (karliss) [inline src/widgets/DebugActions.cpp]: Good fraction of these should just be replaced with `ShortcutManager::makeAction`, afterwards replacing the label if necessary.

File: src/shortcuts/DefaultShortcuts.cpp (hunk 8 above)
Comment (PremadeS) [reply src/shortcuts/DefaultShortcuts.cpp]: > Did you do anything for addressing this?

Yes, replaced `QKeySequence::StandardKey` with `QKeySequence::keyBindings(QKeySequence::StandardKey)`, only where needed

File: src/widgets/BreakpointWidget.cpp (hunk 5 above)
Comment (karliss) [reply src/widgets/BreakpointWidget.cpp]: How about creating one for QShortcut as well (converting to QAction is also an option in some cases). Amount of plain `ShortcutManager::getKeySequence[s]` calls should be kept as low as possible, only for cases that really need custom keyboard event handling. As every call to `::getKeySequences` as opposed to `makeAction`/`setupAction`/`makeQShortCut` will require custom code once the shortcut remapping gets added.

File: src/common/Configuration.cpp [hunk 11]
```diff
@@ -887,6 +888,25 @@ void Configuration::addRecentProject(QString file)
...
+
+void Configuration::setKeySequences(const QString &name, const QList<QKeySequence> &keySequences)
+{
+    QStringList stringList;
+    for (const QKeySequence &seq : keySequences) {
+        stringList.append(seq.toString(QKeySequence::PortableText));
+    }
+    s.setValue("shortcut." + name, stringList);
+}
+
```
Comment (karliss) [inline src/common/Configuration.cpp]: Maybe leave out the actual serialization to settings code from this PR. And just have a placeholder in ShortcutManager for retrieving custom shortcuts which always returns nothing. 

This has potential for it's own share of bikeshedding and also some more serious discussions with regards to handling certain edge cases. It will be a lot easier to discuss and test it once there is GUI for changing them.

File: src/shortcuts/ShortcutManager.cpp (hunk 4 above)
Comment (PremadeS) [reply src/shortcuts/ShortcutManager.cpp]: Should be fixed now hopefully

Comment (karliss) [reply src/shortcuts/ShortcutManager.cpp]: It's fine to leave as is for now, since it will be easy to refactor and exact requirements will be clearer in second part.

But in the PR for shortcut customization empty list probably won't be usable as value for "no user shortcuts available" since it's necessary to distinguish between user not having defined custom shortcut (need to use default) and user having intentionally unmapped the shortcut (because they had no need for it and wanted to use the key for something else).

**Review by karliss — CHANGES_REQUESTED** [review]

File: src/menus/DisassemblyContextMenu.cpp (hunk 9 above)
Comment (PremadeS) [reply src/menus/DisassemblyContextMenu.cpp]: > Can't you use `ShortcutManager::setupAction` here?

Had to change the name of the function because there is another overload of `initAction` with the same parameters.
Let me know if this is what you meant by using `setupAction`

File: src/widgets/BreakpointWidget.cpp (hunk 5 above)
Comment (PremadeS) [reply src/widgets/BreakpointWidget.cpp]: > How about creating one for QShortcut as well (converting to QAction is also an option in some cases). Amount of plain `ShortcutManager::getKeySequence[s]` calls should be kept as low as possible, only for cases that really need custom keyboard event handling. As every call to `::getKeySequences` as opposed to `makeAction`/`setupAction`/`makeQShortCut` will require custom code once the shortcut remapping gets added.

`setupQShortcut()` method doesn't seem useful (for now atleast) hence why it is not implemented. Do correct me if I'm wrong though

File: src/shortcuts/ShortcutManager.cpp [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+QKeySequence ShortcutManager::getKeySequence(const QString &id)
+{
+    const QList<QKeySequence> sequences = getKeySequences(id);
+    return sequences.isEmpty() ? QKeySequence() : sequences.first();
+}
+
+Shortcut ShortcutManager::getShortcut(const QString &id)
+{
+    const auto &defaultShortcuts = getDefaultShortcuts();
+    Shortcut result = defaultShortcuts.value(id);
```
Comment (karliss) [inline src/shortcuts/ShortcutManager.cpp:44]: Could log a warning here in case something is requesting for unrecognized id. The way things are currently it's easy to silently break some shortcuts due to a typo or forgetting to update one of the places during future refactoring of widgets.

File: src/widgets/BreakpointWidget.cpp (hunk 5 above)
Comment (PremadeS) [reply src/widgets/BreakpointWidget.cpp]: > How about creating one for QShortcut as well (converting to QAction is also an option in some cases).

`setupQShortcut()` method doesn't seem useful (for now atleast) hence why it is not implemented. Do correct me if I'm wrong though.

Comment (PremadeS) [reply src/widgets/BreakpointWidget.cpp]: > How about creating one for QShortcut as well (converting to QAction is also an option in some cases).

`setupQShortcut()` method doesn't seem useful (for now atleast) hence why it is not implemented. Do correct me if I'm wrong though. 

Also `QShortcut::setKeys()` (introduced in Qt 6.0) causes qt5 tests to fail, so I should avoid using it right? Currently using `QShortcut::setKey()`

Comment (wargio) [reply src/widgets/BreakpointWidget.cpp]: You can use a macro to check if qt5 or 6 and have a function behave the same but with #if barriers

**Review by karliss — CHANGES_REQUESTED** [review]

File: src/dialogs/preferences/ShortcutOptionsWidget.cpp [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    ui->shortcutTree->expandAll();
+}
+
+void ShortcutOptionsWidget::populateShortcutTree()
+{
+    const QHash<QString, QString> categories = {
+        { "General", tr("General") },
+        { "Disassembly", tr("Disassembly") },
+        { "Decompiler", tr("Decompiler") },
+        { "AddressableItem", tr("AddressableItem") },
```
Comment (karliss) [inline src/dialogs/preferences/ShortcutOptionsWidget.cpp]: "AddressableItem" is acceptable as typeename but not a UI string. If there are no better ideas, at the very least it should be "Addressable items" .

File: src/shortcuts/DefaultShortcuts.cpp [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+          { QKeySequence::keyBindings(QKeySequence::Copy),
+            QT_TRANSLATE_NOOP("DecompilerContextMenu", "Copy"), "DecompilerContextMenu" } },
+        { "Decompiler.copyReferenceAddress",
+          { { Qt::KeyboardModifier::ControlModifier | Qt::KeyboardModifier::ShiftModifier
+              | Qt::Key_C },
+            QT_TRANSLATE_NOOP("DecompilerContextMenu", "Copy address of [flag] (<address>)"),
+            "DecompilerContextMenu" } },
+        { "Decompiler.addComment",
+          { { Qt::Key_Semicolon },
+            QT_TRANSLATE_NOOP("DecompilerContextMenu", "Add Comment"),
```
Comment (karliss) [inline src/shortcuts/DefaultShortcuts.cpp]: This should probably be a common shortcut (in addressable item or general section). I don't think it's necessary to have 5 different "add comment" shortcuts for different widgets and allowing them to be mapped to different key.

File: src/shortcuts/DefaultShortcuts.cpp [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+          { { Qt::Key_Y },
+            QT_TRANSLATE_NOOP("DecompilerContextMenu", "Edit variable <name of variable>"),
+            "DecompilerContextMenu" } },
+        { "Decompiler.toggleBreakpoint",
+          { { Qt::Key_F2, Qt::CTRL | Qt::Key_B },
+            QT_TRANSLATE_NOOP("DecompilerContextMenu", "Add/remove breakpoint"),
+            "DecompilerContextMenu" } },
+        { "Decompiler.advancedBreakpoint",
+          { { Qt::CTRL | Qt::Key_F2 },
+            QT_TRANSLATE_NOOP("DecompilerContextMenu", "Advanced breakpoint"),
```
Comment (karliss) [inline src/shortcuts/DefaultShortcuts.cpp]: F2/Ctrl+F2 should be single shortcut in debug section.

File: src/shortcuts/DefaultShortcuts.cpp [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+          { { Qt::Key_U },
+            QT_TRANSLATE_NOOP("DisassemblyContextMenu", "Undefine function"),
+            "DisassemblyContextMenu" } },
+        { "Disassembly.defineFunction",
+          { { Qt::Key_P },
+            QT_TRANSLATE_NOOP("DisassemblyContextMenu", "Define function here"),
+            "DisassemblyContextMenu" } },
+        { "Disassembly.showXRefs",
+          { { Qt::Key_X },
+            QT_TRANSLATE_NOOP("DisassemblyContextMenu", "Show X-Refs"),
```
Comment (karliss) [inline src/shortcuts/DefaultShortcuts.cpp]: This one also could be common.

File: src/shortcuts/DefaultShortcuts.cpp [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+
+        // General
+        { "General.focusConsole",
+          { { Qt::Key_Period },
+            QT_TRANSLATE_NOOP("MainWindow", "Focus Console Input"),
+            "MainWindow" } },
+        { "General.gotoEntry",
+          { { Qt::Key_G }, QT_TRANSLATE_NOOP("MainWindow", "Go to Entry"), "MainWindow" } },
+        { "General.seek",
+          { { Qt::Key_S }, QT_TRANSLATE_NOOP("MainWindow", "Seek"), "MainWindow" } },
```
Comment (karliss) [inline src/shortcuts/DefaultShortcuts.cpp]: Two differently named shortcuts for the same action. 

In other situations I would suggest to just assign two different shortcut sequences for same shortcut action (if we really needed multiple shortcuts).

But the single letter "G" is problematic and conflicts with disassembly global variable shortcut.  Sou I would suggest just getting rid of this "G".

File: src/shortcuts/DefaultShortcuts.cpp [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+          { { Qt::Key_Escape },
+            QT_TRANSLATE_NOOP("DisassemblyWidget", "Seek to Previous Address"),
+            "DisassemblyWidget" } },
+        { "Disassembly.moveDownJ",
+          { { Qt::Key_J },
+            QT_TRANSLATE_NOOP("DisassemblyWidget", "Move Cursor Down"),
+            "DisassemblyWidget" } },
+        { "Disassembly.moveDown",
+          { QKeySequence::keyBindings(QKeySequence::MoveToNextLine),
+            QT_TRANSLATE_NOOP("DisassemblyWidget", "Move Cursor Down"), "DisassemblyWidget" } },
```
Comment (karliss) [inline src/shortcuts/DefaultShortcuts.cpp]: Shouldn't these be two shortcut sequences in single shortcut entry?

File: src/shortcuts/DefaultShortcuts.cpp [hunk 19]
```diff
@@ -0,0 +1,348 @@
...
+            "DisassemblyContextMenu" } },
+        { "Disassembly.setAsString",
+          { { Qt::Key_A },
+            QT_TRANSLATE_NOOP("DisassemblyContextMenu", "Auto-detect"),
+            "DisassemblyContextMenu" } },
+        { "Disassembly.setAsStringAdvanced",
+          { { Qt::SHIFT | Qt::Key_A },
+            QT_TRANSLATE_NOOP("DisassemblyContextMenu", "Advanced"),
+            "DisassemblyContextMenu" } },
+        { "Disassembly.setToDataEx", { { Qt::Key_Asterisk }, "...", "DisassemblyContextMenu" } },
```
Comment (karliss) [inline src/shortcuts/DefaultShortcuts.cpp]: Needs a proper label, out of context the "..." label is confusing.

Overriding it in the corresponding context menu if it's not possible to come with something short that suits both.

File: src/shortcuts/DefaultShortcuts.cpp [hunk 20]
```diff
@@ -0,0 +1,348 @@
...
+            QT_TRANSLATE_NOOP("DisassemblyContextMenu", "Show X-Refs"),
+            "DisassemblyContextMenu" } },
+        { "Disassembly.XRefsForVariables",
+          { { QKeySequence(Qt::SHIFT | Qt::Key_X) },
+            QT_TRANSLATE_NOOP("DisassemblyContextMenu", "X-Refs for local variables"),
+            "DisassemblyContextMenu" } },
+        { "Disassembly.showOptions",
+          { { Qt::Key_D },
+            QT_TRANSLATE_NOOP("DisassemblyContextMenu", "Show Options"),
+            "DisassemblyContextMenu" } },
```
Comment (karliss) [inline src/shortcuts/DefaultShortcuts.cpp]: What is this? Need to look into more carefully. 

Couldn't find  any "show options" action the disassembly context menu. Looking at the code made this look even weirder. Supposedly it actually refers to the main Cutter options dialog. But why "D" and why in disassembly context menu. Also it wasn't immediately obvious why it doesn't show up in disassembly context menu.

File: src/widgets/OverviewWidget.cpp [hunk 21]
```diff
@@ -17,10 +18,10 @@ OverviewWidget::OverviewWidget(MainWindow *main) : CutterDockWidget(main)
     graphDataRefreshDeferrer = createRefreshDeferrer([this]() { updateGraphData(); });
 
     // Zoom shortcuts
-    QShortcut *shortcut_zoom_in = new QShortcut(QKeySequence(Qt::Key_Plus), this);
+    QShortcut *shortcut_zoom_in = Shortcuts()->makeQShortcut("General.zoomIn", this);
```
Comment (karliss) [inline src/widgets/OverviewWidget.cpp]: These are graph overview widget specific zoom in/zoom out actions. 

Different from the global "Ctrl++" "Ctrl--" global zoom shortcuts that change the font size multiplier in disassembly/graph and few other widget. And it seems that the "Ctrl++" doesn't show up in the new list at all.

File: src/menus/DisassemblyContextMenu.cpp [hunk 22]
```diff
@@ -71,32 +72,32 @@ DisassemblyContextMenu::DisassemblyContextMenu(QWidget *parent, MainWindow *main
...
       showInSubmenu(this)
 {
-    initAction(&actionCopy, tr("Copy"), SLOT(on_actionCopy_triggered()), getCopySequence());
+    initShortcutAction(&actionCopy, "Disassembly.copy", SLOT(on_actionCopy_triggered()));
     addAction(&actionCopy);
 
-    initAction(&actionCopyAddr, tr("Copy address"), SLOT(on_actionCopyAddr_triggered()),
-               getCopyAddressSequence());
+    initShortcutAction(&actionCopyAddr, "Disassembly.copyAddress",
+                       SLOT(on_actionCopyAddr_triggered()));
```
Comment (karliss) [inline src/menus/DisassemblyContextMenu.cpp]: This should also be a single unified shortcut instead of unique per/widget.

**Review by karliss — CHANGES_REQUESTED** [review]

File: src/shortcuts/DefaultShortcuts.cpp (hunk 16 above)
Comment (PremadeS) [reply src/shortcuts/DefaultShortcuts.cpp]: > This one also could be common.

what should be used as the context string for this? because there are multiple places that have the translations (`DisassemblyContextMeny` , `AddressableItem` ...) should I just pick one?

File: src/shortcuts/DefaultShortcuts.cpp (hunk 20 above)
Comment (PremadeS) [reply src/shortcuts/DefaultShortcuts.cpp]: > What is this? Need to look into more carefully.
> 
> Couldn't find any "show options" action the disassembly context menu. Looking at the code made this look even weirder. Supposedly it actually refers to the main Cutter options dialog. But why "D" and why in disassembly context menu. Also it wasn't immediately obvious why it doesn't show up in disassembly context menu.

call to `addAction()` was missing that's why it wasn't showing up.

It supposedly shows the preferences options dialog for disassembly but due to wrong ordering of `PreferencesDialog::Section` **enum**  in `PreferencesDialog.h`:
```
enum class Section { Appearance, Disassembly }; // wrong order + contains only 2 entries instead of the usual 6, excluding Shortcuts
```
 it shows "Debug" options instead of "Disassembly" because Debug is the second entry in the original categories list.

Closing the preference dialog crashes the application. Not immediately obvious why

Also the "D" was added by me since the original code had this:
```
QKeySequence DisassemblyContextMenu::getDisplayOptionsSequence() const
{
    return {}; // TODO insert correct sequence
}
```
In my opinion it doesn't seem useful to have this

File: src/shortcuts/DefaultShortcuts.cpp (hunk 16 above)
Comment (karliss) [reply src/shortcuts/DefaultShortcuts.cpp]: Yes just pick one. 

If a widget needs widget specific variant of label, then the override can go in the corresponding file with it's context.

File: src/shortcuts/DefaultShortcuts.cpp (hunk 20 above)
Comment (karliss) [reply src/shortcuts/DefaultShortcuts.cpp]: Agree,  remove it. The disassembly context menu is already gigantic no need to put even more stuff there.

File: src/shortcuts/DefaultShortcuts.cpp [hunk 23]
```diff
@@ -0,0 +1,299 @@
...
+
+const QHash<QString, Shortcut> &getDefaultShortcuts()
+{
+    static const QHash<QString, Shortcut> defaultShortcuts = {
+
+        // General
+        { "General.focusConsole",
+          { { Qt::Key_Period },
+            QT_TRANSLATE_NOOP("MainWindow", "Focus Console Input"),
+            "MainWindow" } },
```
Comment (karliss) [inline src/shortcuts/DefaultShortcuts.cpp]: How about placing this in console section.

File: src/shortcuts/DefaultShortcuts.cpp [hunk 24]
```diff
@@ -0,0 +1,299 @@
...
+          { { Qt::Key_X },
+            QT_TRANSLATE_NOOP("AddressableItemContextMenu", "Show X-Refs"),
+            "AddressableItemContextMenu" } },
+        { "General.copyAddress",
+          { { Qt::CTRL | Qt::SHIFT | Qt::Key_C },
+            QT_TRANSLATE_NOOP("AddressableItemContextMenu", "Copy address"),
+            "AddressableItemContextMenu" } },
+        { "General.copy",
+          { QKeySequence::keyBindings(QKeySequence::Copy),
+            QT_TRANSLATE_NOOP("DecompilerContextMenu", "Copy"), "DecompilerContextMenu" } },
```
Comment (karliss) [inline src/shortcuts/DefaultShortcuts.cpp]: Copy is one thing I intentionally didn't mention when discussing making common shortcuts. 

Having it copy with "Ctrl+c" in general section will likely give a false impression, that overriding it will affect a lot more than it does. While in reality it only affects few custom actions in specific widgets.

File: src/shortcuts/ShortcutManager.cpp (hunk 12 above)
Comment (karliss) [reply src/shortcuts/ShortcutManager.cpp:44]: Wanted to confirm this is working and all the ids are correct. Modified one of the shortcut ids and it didn't cause a warning.

Some paths like getKeySequences, and makeQShortcut don't use this, leading to potential silent failures.

File: src/widgets/CutterGraphView.cpp [hunk 25]
```diff
@@ -212,8 +207,9 @@ bool CutterGraphView::event(QEvent *event)
     case QEvent::ShortcutOverride: {
         QKeyEvent *keyEvent = static_cast<QKeyEvent *>(event);
         qhelpers::KeyComb key = Qt::Key(keyEvent->key()) | keyEvent->modifiers();
-        if (key == KEY_ZOOM_OUT || key == KEY_ZOOM_RESET || key == KEY_ZOOM_IN
-            || key == KEY_ZOOM_IN2) {
+        if (Shortcuts()->matchesKeySequence("General.zoomOut", key)
+            || Shortcuts()->matchesKeySequence("General.zoomReset", key)
+            || Shortcuts()->matchesKeySequence("General.zoomIn", key)) {
```
Comment (karliss) [inline src/widgets/CutterGraphView.cpp:212]: The way `matchesKeySequence` currently works makes me slightly uncomfortable. You are making a copies of a bunch of nontrivial objects on every key press. Not the end of world and can be probably left as is, but not ideal.

**Review by karliss — CHANGES_REQUESTED** [review]
Hopefully last round of changes.

File: src/widgets/CutterGraphView.cpp (hunk 25 above)
Comment (PremadeS) [reply src/widgets/CutterGraphView.cpp:212]: Unfortunately wasn't able to come up with something better.

An alternate approach could be to avoid using the `getkeySequences()` in `matchesKeySequence()` and check for custom key sequences inside of `matchesKeySequence()`. But that would duplicate logic in two different places which doesn't sound ideal.

If this should be handled in a specific way please do let me know and I will change it.

**Review by karliss — APPROVED** [review]
